README.md
# nautilus-advacl
A nautilus plugin to manage advanced ACLs

![alt text](https://hadler.me/wordpress/wp-content/uploads/2018/04/advacl1.png "View of Nautilus property page")


debian/changelog
nautilus-advacl (0.8-2) precise; urgency=low

  * Added dependency for python-pyinotify

 -- Rene Hadler <rene@iksdeh.at>  Fri, 18 Jul 2014 10:06:07 +0100

nautilus-advacl (0.8-1) precise; urgency=low

  * Package does not overwrite nautilus language anymore

 -- Rene Hadler <rene@iksdeh.at>  Sun, 31 Mar 2013 14:06:07 +0100

nautilus-advacl (0.7-1) precise; urgency=low

  * Code & GUI cleanup
  * Added icons to some views
  * Language files added

 -- Rene Hadler <rene@iksdeh.at>  Sun, 10 Feb 2013 17:56:07 +0100

nautilus-advacl (0.6-1) precise; urgency=low

  * Using polkit instead of sudo
  * pyinotify is not watching the file/directory recursive

 -- Rene Hadler <rene@iksdeh.at>  Sat, 09 Feb 2013 20:01:00 +0100

nautilus-advacl (0.5) precise; urgency=low

  * initial release

 -- Rene Hadler <rene@iksdeh.at> Sat, 02 Feb 2013 14:26:32 +0100


test.py
'''
Created on 19.01.2013

@author: rene
'''

#!/usr/bin/python2

import sys
import os

from gi.repository import Nautilus, GObject, Gtk
#

sys.path.append(os.path.dirname(os.path.realpath(__file__)) + "/nautilus-advacl")

import nautilusadvacllib
from nautiluspropaddacl import NautilusWindowAddACL

print sys.path

def tvObjects_sel_changed(sel):
    #print "selection changed2!!!"
    global tvObjects

    tv, iter = sel.get_selected()
    if not iter:
        return
    
    model = tvObjects.get_model()
    objACL = model.get_value(iter, 0)
    #print "selected", model.get_value(iter, 1)
    tvPermissions_set_permision(objACL.perm)
    
def tvPermissions_set_permision(objPerm):
    global tvPermissions
    
    model = tvPermissions.get_model()
    
    model[0][1] = objPerm.read
    model[1][1] = objPerm.write
    model[2][1] = objPerm.execute
    
def on_cell_toggled(widget, path):
    global tvPermissions
    
    model = tvPermissions.get_model()
    iter = model.get_iter(path)
    state = model.get_value(iter, 1)
    if state == True:
        model.set_value(iter, 1, False)
    elif state == False:
        model.set_value(iter, 1, True)
        
def btnObjAdd_clicked(button):
    builder.add_from_file("/home/rene/DEV/eclipse/nautilus-advacl/nautilus-advacl/nautilus-prop-add-acl.glade")
    #bbox = builder.get_objects()
    bbox = builder.get_object("boxMain")
    win_add_acl = NautilusWindowAddACL()
    win_add_acl.set_modal(True)
    win_add_acl.add(bbox)
    win_add_acl.show()

builder = Gtk.Builder()
#builder.add_objects_from_file("/home/rene/DEV/eclipse/nautilus-advacl/nautilus-prop.glade", ["boxMain"])
builder.add_from_file("/home/rene/DEV/eclipse/nautilus-advacl/nautilus-advacl/nautilus-prop.glade")
#bbox = builder.get_objects()
bbox = builder.get_object("window1")
bbox.connect("destroy", Gtk.main_quit)
bbox.set_position(Gtk.WindowPosition.CENTER)
bbox.show()

# Treeview
#store = Gtk.ListStore(str)
#store.append(["test1"])
#store.append(["test2"])
#store.append(["test3"])

tvObjects = builder.get_object("tvObjects")
#tvObjects.set_model(store)

renderer = Gtk.CellRendererText()
column = Gtk.TreeViewColumn("Objekt", renderer, text=1)
tvObjects.append_column(column)

selection = tvObjects.get_selection()
selection.connect("changed", tvObjects_sel_changed)

# Treeview2
store2 = Gtk.ListStore(str, bool)
store2.append(["Lesen", False])
store2.append(["Schreiben", False])
store2.append(["Ausfuehren", False])

tvPermissions = builder.get_object("tvPermissions")
tvPermissions.set_model(store2)

renderer2 = Gtk.CellRendererText()
column2 = Gtk.TreeViewColumn("Objekt", renderer, text=0)
column2.set_min_width(250)
tvPermissions.append_column(column2)

renderer_toggle = Gtk.CellRendererToggle()
renderer_toggle.connect("toggled", on_cell_toggled)
column_toggle = Gtk.TreeViewColumn("Zulassen", renderer_toggle, active=1)
tvPermissions.append_column(column_toggle)

#renderer_toggle2 = Gtk.CellRendererToggle()
#column_toggle2 = Gtk.TreeViewColumn("Verweigern", renderer_toggle2, active=2)
#tvPermissions.append_column(column_toggle2)

btnObjAdd = builder.get_object("btnObjAdd")
btnObjAdd.connect("clicked", btnObjAdd_clicked)

lib = nautilusadvacllib.AdvACLLibrary()
perms = lib.get_permissions("/home/rene/tmp/test")

store = Gtk.ListStore(GObject.TYPE_PYOBJECT, str)

for perm in perms:
    print perm.realm, perm.object, perm.perm
    store.append([perm, perm.object])
    
tvObjects.set_model(store)

Gtk.main()

nautilus-advacl/nautilusadvacllib.py
'''
Created on 25.01.2013

@author: Rene Hadler
@email:  rene@iksdeh.at
'''

import os
import re
import subprocess

class AdcACLPermission:
    def __init__(self, perm=None):
        self.read = False
        self.write = False
        self.execute = False
        self.changed = False
        
        self.convert(perm)
        
    def convert(self, perm):
        if perm == None:
            return
        
        if perm[0] == "r":
            self.read = True
            
        if perm[1] == "w":
            self.write = True
            
        if perm[2] == "x":
            self.execute = True
            
    def setPerm(self, attr, state):
        self.changed = True
        
        if attr == "r":
            self.read = state
        elif attr == "w":
            self.write = state
        elif attr == "x":
            self.execute = state
            
    def format_as_string(self):
        strPerm = ""
        
        strPerm += "r" if self.read else "-"
        strPerm += "w" if self.write else "-"
        strPerm += "x" if self.execute else "-"
        
        return strPerm
            
class AdcACLObject:
    def __init__(self, a_realm, a_object, a_perm=None, a_default=False):
        self.realm = a_realm
        self.object = a_object
        self.default = a_default
        if a_perm == None:
            self.perm = None
        else:
            self.perm = AdcACLPermission(a_perm)

class AdvACLLibrary:
    def __init__(self):
        self.re_stdacl = re.compile("^(user|group|mask|other):([^:]*):([rwx\-]{3})")
        self.re_dflacl = re.compile("^default:(user|group|mask|other):([^:]*):([rwx\-]{3})")
    
    def get_permissions(self, filename, a_default=False):
        perm = []
        
        if not os.path.exists(filename):
            return perm
        
        output = subprocess.check_output(["getfacl", filename])
        out_lines = output.split("\n")
        
        regexp = ""
        if a_default == True:
            regexp = self.re_dflacl
        else:
            regexp = self.re_stdacl
        
        for line in out_lines:
            m_result = regexp.match(line)
            if m_result:
                res_realm = m_result.group(1)
                res_object = m_result.group(2)
                res_perm = m_result.group(3)
                
                if not res_object:
                    continue
                
                perm.append(AdcACLObject(res_realm, res_object, res_perm, a_default))
                    
        return perm
    
    def set_permissions(self, objAdcACL, filename):
        strPerm = ""
        
        if objAdcACL.default == True:
            strPerm += "default:"
        
        strPerm += objAdcACL.realm + ":"
        strPerm += objAdcACL.object + ":"
        strPerm += objAdcACL.perm.format_as_string() + " "
        
        try:
            subprocess.check_output(["setfacl", "-m", strPerm, filename])
        except subprocess.CalledProcessError:
            print "No success running as user, try with sudo and then give up..."
            try:
                subprocess.check_output(["pkexec", "setfacl", "-m", strPerm, filename])
            except subprocess.CalledProcessError:
                print "Command denied, coninue..."
            
    def remove_acl(self, objAdcACL, filename):
        strRemove = ""
        
        if objAdcACL.default == True:
            strRemove += "default:"
        
        strRemove += objAdcACL.realm + ":"
        strRemove += objAdcACL.object
        
        try:
            subprocess.check_output(["setfacl", "-x", strRemove, filename])
            return True
        except subprocess.CalledProcessError as e:
            print "No success running as user, try with sudo and then give up..."
            try:
                subprocess.check_output(["pkexec", "setfacl", "-x", strRemove, filename])
                return True
            except subprocess.CalledProcessError:
                print "Command denied, coninue..."
                return False

nautilus-advacl/nautiluspropaddacl.py
'''
Created on 25.01.2013

@author: Rene Hadler
@email:  rene@iksdeh.at
'''

import os
import sys
import locale
import urllib
import gettext
_ = gettext.gettext

import pwd, grp

from gi.repository import Nautilus, GObject, Gtk, GdkPixbuf

WORK_DIR = os.path.dirname(os.path.realpath(__file__))
sys.path.append(WORK_DIR + "/..")

import nautilusadvacllib as advacllib

class NautilusWindowAddACL(Gtk.Window):
    def __init__(self, window):
        Gtk.Window.__init__(self)
        self.advacllibrary = advacllib.AdvACLLibrary()
        
        self.set_title(_("Nautilus - Add new ACL"))
        self.set_position(Gtk.WindowPosition.CENTER)
        self.set_default_size(480, 600)
        self.set_modal(True)
        
        self.objWindowMain = window
        
        self.init_widgets()
        
    def init_widgets(self):
        print "init widgets"
        self.boxMain = self.objWindowMain.builder_add_acl.get_object("boxMain")
        self.objTypes = self.objWindowMain.builder_add_acl.get_object("cbObjectTypes")
        self.tvObjects = self.objWindowMain.builder_add_acl.get_object("tvObjects")
        self.tvPermissions = self.objWindowMain.builder_add_acl.get_object("tvPermissions")
        self.btnAddObject = self.objWindowMain.builder_add_acl.get_object("btnAddObject")
        self.btnCancel = self.objWindowMain.builder_add_acl.get_object("btnCancel")
        self.cbxDefaultACL = self.objWindowMain.builder_add_acl.get_object("cbxDefaultACL")
        
        self.add(self.boxMain)
        
        # tvPermissions
        renderer2 = Gtk.CellRendererText()
        column2 = Gtk.TreeViewColumn(_("Object"), renderer2, text=1)
        column2.set_min_width(350)
        self.tvPermissions.append_column(column2)
        
        renderer_toggle = Gtk.CellRendererToggle()
        renderer_toggle.connect("toggled", self.tvPermissions_toggled)
        column_toggle = Gtk.TreeViewColumn(_("Grant"), renderer_toggle, active=2)
        column_toggle.set_min_width(100)
        self.tvPermissions.append_column(column_toggle)
        
        # tvPermissions data
        store = Gtk.ListStore(str, str, bool)
        store.append(["r", _("Read"), False])
        store.append(["w", _("Write"), False])
        store.append(["x", _("Execute"), False])
        self.tvPermissions.set_model(store)
        
        # objTypes
        self.icon_user = GdkPixbuf.Pixbuf.new_from_file_at_size(WORK_DIR + '/img/icon-user.png', 16, 16)
        self.icon_group = GdkPixbuf.Pixbuf.new_from_file_at_size(WORK_DIR + '/img/icon-group.png', 16, 16)
        
        objTypesStore = Gtk.ListStore(str, str, GdkPixbuf.Pixbuf)
        objTypesStore.append(["user", _("Users"), self.icon_user])
        objTypesStore.append(["group", _("Groups"), self.icon_group])
        self.objTypes.set_model(objTypesStore)
        
        renderer_text = Gtk.CellRendererText()
        renderer_pixbuf = Gtk.CellRendererPixbuf()
        self.objTypes.connect("changed", self.cbObjectTypes_changed)
        self.objTypes.pack_start(renderer_pixbuf, False)
        self.objTypes.add_attribute(renderer_pixbuf, "pixbuf", 2)
        self.objTypes.pack_start(renderer_text, True)
        self.objTypes.add_attribute(renderer_text, "text", 1)
        self.objTypes.set_active(0)
        
        # tvObjects
        renderer = Gtk.CellRendererText()
        renderer_pixbuf2 = Gtk.CellRendererPixbuf()
        column = Gtk.TreeViewColumn(_("Object"))
        column.pack_start(renderer_pixbuf2, False)
        column.add_attribute(renderer_pixbuf2, "pixbuf", 2)
        column.pack_start(renderer, True)
        column.add_attribute(renderer, "text", 1)
        column.set_sort_column_id(1)
        self.tvObjects.append_column(column)
        
        # btnObjAdd
        self.btnAddObject.connect("clicked", self.btnAddObject_clicked)
        
        # btnCancel
        self.btnCancel.connect("clicked", self.btnCancel_clicked)
        
        # Do default sorting when init widgets
        column.clicked()
        
    def cbObjectTypes_changed(self, combo):
        # Type has changed, we have to update our list of acl objects
        model = self.objTypes.get_model()
        type = model.get_value(combo.get_active_iter(), 0)
        self.tvObjects.set_model(None)
        
        objStore = Gtk.ListStore(str, str, GdkPixbuf.Pixbuf)
        
        if type == "group":
            groups = grp.getgrall()
            for group in groups:
                objStore.append([group[0], group[0], self.icon_group])
                
        elif type == "user":
            users = pwd.getpwall()
            for user in users:
                objStore.append([user[0], user[0], self.icon_user])
                
        self.tvObjects.set_model(objStore)
        
        # Do default sorting when reloading list
        column = self.tvObjects.get_column(0)
        if column:
            column.clicked()
        
    def tvPermissions_toggled(self, widget, path):
        model = self.tvPermissions.get_model()
        iter = model.get_iter(path)
        attr = model.get_value(iter, 0)
        state = model.get_value(iter, 2)
        
        if state == True:
            model.set_value(iter, 2, False)
        elif state == False:
            model.set_value(iter, 2, True)
            
    def btnAddObject_clicked(self, button):
        model = self.tvPermissions.get_model()
        selection = self.tvObjects.get_selection()
        tv, iterObjects = selection.get_selected()
        if iterObjects == None:
            dialog = Gtk.MessageDialog(self, 0, Gtk.MessageType.INFO, Gtk.ButtonsType.OK, _("Not all required fields was provided"))
            dialog.format_secondary_text(_("Please choose a object you want to add."))
            dialog.run()    
            dialog.destroy()
        
            return 
        
        # Get ACL object
        objectsModel = self.tvObjects.get_model()
        object = objectsModel.get_value(iterObjects, 0)
        
        # Get ACL type
        idx = self.objTypes.get_active()
        type = self.objTypes.get_model()[idx][0]
        
        # Get ACL permissions
        permModel = self.tvPermissions.get_model()
        
        objAdvACL = advacllib.AdcACLObject(type, object, a_default=self.cbxDefaultACL.get_active())
        objPerm = advacllib.AdcACLPermission()
        objPerm.setPerm(permModel[0][0], permModel[0][2])
        objPerm.setPerm(permModel[1][0], permModel[1][2])
        objPerm.setPerm(permModel[2][0], permModel[2][2])
        objAdvACL.perm = objPerm
        
        self.advacllibrary.set_permissions(objAdvACL, self.objWindowMain.filename)
        
        #self.objWindowMain.load_acls()
        self.destroy()
        
    def btnCancel_clicked(self, button):
        self.destroy()

nautilus-advacl-properties.py
'''
Created on 20.01.2013

@author: Rene Hadler
@email:  rene@iksdeh.at
'''

import os
import sys
import locale
import urllib
import pyinotify
import gettext
_ = gettext.gettext

from gi.repository import Nautilus, GObject, Gtk, GLib, GdkPixbuf

WORK_DIR = os.path.dirname(os.path.realpath(__file__))
sys.path.append(WORK_DIR + "/nautilus-advacl")

import nautilusadvacllib as advacllib
from nautiluspropaddacl import NautilusWindowAddACL

class FileEvent(pyinotify.ProcessEvent):
    def __init__(self, objMain):
        self.main = objMain
    
    def process_IN_ATTRIB(self, event):
        print "Metadata changed:", event.pathname
        self.main.load_acls()
        self.main.tvObjects_sel_first_row()

class AdvACLExtension(GObject.GObject, Nautilus.PropertyPageProvider):
    def __init__(self):
        #gettext.bindtextdomain('nautilusadvacl', WORK_DIR + '/nautilus-advacl/locale/')
        gettext.textdomain('nautilusadvacl')
        self.advacllibrary = advacllib.AdvACLLibrary()
    
    def init_fileeventhandler(self):
        # Install inotify handler
        print "Observing: ", self.filename
        
        wm = pyinotify.WatchManager()
        mask = pyinotify.IN_ATTRIB
        handler = FileEvent(self)
        self.notifierFile = pyinotify.Notifier(wm, handler, timeout=10)
        wdd = wm.add_watch(self.filename, mask, rec=False)

        GLib.timeout_add(500, self.checkFileEvents)
        
    def checkFileEvents(self):
        self.notifierFile.process_events()
        while self.notifierFile.check_events():
            self.notifierFile.read_events()
            self.notifierFile.process_events()
            
        return True
    
    def setColumnOrder(self):
        # Do default sorting when init
        column = self.tvObjects.get_column(0)
        if column:
            column.clicked()
            
        self.tvObjects_sel_first_row()
        
        return False
    
    def get_property_pages(self, files):
        if len(files) != 1:
            return
        
        file = files[0]
        print file.get_uri_scheme()
        if file.get_uri_scheme() != 'file':
            return
        
        self.filename = urllib.unquote(file.get_uri()[7:])
        self.init_fileeventhandler()

        self.property_label = Gtk.Label(_("Advanced ACL"))
        self.property_label.show()   
        
        self.builder = Gtk.Builder()
        self.builder.set_translation_domain('nautilusadvacl')
        self.builder.add_objects_from_file(WORK_DIR + "/nautilus-advacl/nautilus-prop.glade", ["boxMain"])
        self.bbox = self.builder.get_object("boxMain")
        self.bbox.show()
        
        self.init_widgets()
        self.load_acls()
        
        return Nautilus.PropertyPage(name="Advanced ACL",
                                     label=self.property_label, 
                                     page=self.bbox),
                                    
    def load_acls(self):
        # We load the acls from file and update the treeview
        tvObjects = self.builder.get_object("tvObjects")
        tvPermissions = self.builder.get_object("tvPermissions")
        tvObjects.set_model(None)
        
        icon_user = GdkPixbuf.Pixbuf.new_from_file_at_size(WORK_DIR + '/nautilus-advacl/img/icon-user.png', 16, 16)
        icon_group = GdkPixbuf.Pixbuf.new_from_file_at_size(WORK_DIR + '/nautilus-advacl/img/icon-group.png', 16, 16)
        
        permList = self.advacllibrary.get_permissions(self.filename, self.cbxDefaultACL.get_active())
        permStore = Gtk.ListStore(GObject.TYPE_PYOBJECT, str, GdkPixbuf.Pixbuf)
        
        for permObj in permList:
            icon = icon_user if permObj.realm == "user" else icon_group
            permStore.append([permObj, permObj.object + " (" + permObj.realm + ")", icon])
            
        tvObjects.set_model(permStore)
        self.setColumnOrder()
        
    def init_widgets(self):
        self.tvObjects = self.builder.get_object("tvObjects")
        self.tvPermissions = self.builder.get_object("tvPermissions")
        self.btnPermApply = self.builder.get_object("btnPermApply")
        self.btnObjRemove = self.builder.get_object("btnObjRemove")
        self.btnObjAdd = self.builder.get_object("btnObjAdd")
        self.cbxDefaultACL = self.builder.get_object("cbxDefaultACL")
        
        # tvObjects
        renderer = Gtk.CellRendererText()
        renderer_pixbuf = Gtk.CellRendererPixbuf()
        column = Gtk.TreeViewColumn(_("Object"))
        column.pack_start(renderer_pixbuf, False)
        column.add_attribute(renderer_pixbuf, "pixbuf", 2)
        column.pack_start(renderer, True)
        column.add_attribute(renderer, "text", 1)
        column.set_sort_column_id(1)
        self.tvObjects.append_column(column)
        
        selection = self.tvObjects.get_selection()
        selection.connect("changed", self.tvObjects_sel_changed)
        
        # tvPermissions
        renderer2 = Gtk.CellRendererText()
        column2 = Gtk.TreeViewColumn(_("Object"), renderer2, text=1)
        column2.set_min_width(350)
        self.tvPermissions.append_column(column2)
        
        renderer_toggle = Gtk.CellRendererToggle()
        renderer_toggle.connect("toggled", self.tvPermissions_toggled)
        column_toggle = Gtk.TreeViewColumn(_("Grant"), renderer_toggle, active=2)
        column_toggle.set_min_width(100)
        self.tvPermissions.append_column(column_toggle)
        
        # tvPermissions data
        store = Gtk.ListStore(str, str, bool)
        store.append(["r", _("Read"), False])
        store.append(["w", _("Write"), False])
        store.append(["x", _("Execute"), False])
        self.tvPermissions.set_model(store)
        
        # btnPermApply
        self.btnPermApply.connect("clicked", self.btnPermApply_clicked)
        
        # btnObjRemove
        self.btnObjRemove.connect("clicked", self.btnObjRemove_clicked)
        
        # btnObjAdd
        self.btnObjAdd.connect("clicked", self.btnObjAdd_clicked)

        # on_destroy event
        self.btnObjAdd.connect("destroy", self.cleanupAdvACL)
        
        self.cbxDefaultACL.connect("toggled", self.cbxDefaultACL_toggled)
        
    def tvObjects_sel_first_row(self):
        model = self.tvObjects.get_model()
        if model.get_iter_first() == None:
            return
        
        selection = self.tvObjects.get_selection()
        selection.select_iter(model.get_iter_first())
        
    def tvObjects_sel_changed(self, sel):
        #print "selection changed2!!!"

        tv, iter = sel.get_selected()
        if not iter:
            return
        
        model = self.tvObjects.get_model()
        objACL = model.get_value(iter, 0)
        self.tvPermissions_set_permission(objACL.perm)
        
    def tvPermissions_set_permission(self, objPerm):
        model = self.tvPermissions.get_model()
        
        model[0][2] = objPerm.read
        model[1][2] = objPerm.write
        model[2][2] = objPerm.execute
        
    def tvPermissions_toggled(self, widget, path):
        model = self.tvPermissions.get_model()
        iter = model.get_iter(path)
        attr = model.get_value(iter, 0)
        state = model.get_value(iter, 2)
        
        selection = self.tvObjects.get_selection()
        tv, iterObjects = selection.get_selected()
        objectsModel = self.tvObjects.get_model()
        objACL = objectsModel.get_value(iterObjects, 0)
        
        print objACL.perm.format_as_string()
        
        if state == True:
            model.set_value(iter, 2, False)
            objACL.perm.setPerm(attr, False)
        elif state == False:
            model.set_value(iter, 2, True)
            objACL.perm.setPerm(attr, True)
            
    def btnPermApply_clicked(self, button):
        objectsModel = self.tvObjects.get_model()
        
        # We check now if there are any changed settings
        for aclObj in objectsModel:
            objACL = aclObj[0]
            if objACL.perm.changed == True:
                print objACL.object
                self.advacllibrary.set_permissions(objACL, self.filename)
                
    def btnObjRemove_clicked(self, button):
        objectsModel = self.tvObjects.get_model()
        selection = self.tvObjects.get_selection()
        tv, iterObjects = selection.get_selected()
        if not iterObjects:
            return
        
        objACL = objectsModel.get_value(iterObjects, 0)
        
        # We remove now the selected entry
        print "removing", objACL.object
        if self.advacllibrary.remove_acl(objACL, self.filename):
            objectsModel.remove(iterObjects)
            
    def btnObjAdd_clicked(self, button):
        # Yeah, we need to load the glade file everytime coz else it will only show up the first time
        self.builder_add_acl = Gtk.Builder()
        self.builder_add_acl.set_translation_domain('nautilusadvacl')
        self.builder_add_acl.add_objects_from_file(WORK_DIR + "/nautilus-advacl/nautilus-prop-add-acl.glade", ["boxMain"])
        self.win_add_acl = NautilusWindowAddACL(self)
        self.win_add_acl.show()
        
    def cleanupAdvACL(self, data):
        print "cleanup"
        
    def cbxDefaultACL_toggled(self, cbx):
        print "default toggled"
        self.load_acls()
        self.tvObjects_sel_first_row()